src/shims/unix/linux/epoll.rs

```diff
@@ -168,25 +243,47 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
     fn epoll_wait(
         &mut self,
         epfd: &OpTy<'tcx>,
-        events: &OpTy<'tcx>,
+        events_op: &OpTy<'tcx>,
         maxevents: &OpTy<'tcx>,
         timeout: &OpTy<'tcx>,
     ) -> InterpResult<'tcx, Scalar> {
         let this = self.eval_context_mut();
 
         let epfd = this.read_scalar(epfd)?.to_i32()?;
-        let _events = this.read_scalar(events)?.to_pointer(this)?;
-        let _maxevents = this.read_scalar(maxevents)?.to_i32()?;
+        let maxevents = this.read_scalar(maxevents)?.to_i32()?;
+        let event = this.deref_pointer_as(
+            events_op,
+            this.libc_array_ty_layout("epoll_event", maxevents.try_into().unwrap()),
+        )?;
         let _timeout = this.read_scalar(timeout)?.to_i32()?;
 
         let Some(mut epfd) = this.machine.fds.get_mut(epfd) else {
             return Ok(Scalar::from_i32(this.fd_not_found()?));
         };
-        let _epfd = epfd
+        let epoll_file_description = epfd
             .downcast_mut::<Epoll>()
             .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_wait`"))?;
 
-        // FIXME return number of events ready when scheme for marking events ready exists
-        throw_unsup_format!("returning ready events from epoll_wait is not yet implemented");
+        let binding = epoll_file_description.return_ready_list()?;
+        let ready_list = binding.borrow_mut();
+        let mut _num_of_events: i32 = 0;
+        let mut array_iter = this.project_array_fields(&event)?;
+        drop(epfd);
+        for (epoll_key, _epoll_return) in ready_list.iter() {
+            if let Some(_file_description) = epoll_key.0.get_file_description() {
+                // Return notification to the caller if the file description is not fully closed.
+                let src =
+                    this.allocate_epoll_event(_epoll_return.events, _epoll_return.data).unwrap();
+                let des = array_iter.next(this)?.unwrap().1;
+                this.copy_op(&src, &des)?;
```

Comment:
Write directly to fields of this place instead of allocating a copy and copying it over